crates/cext/src/dag.rs

```diff
@@ -706,6 +708,96 @@ pub unsafe extern "C" fn qk_dag_op_node_kind(dag: *const DAGCircuit, node: u32)
     }
 }
 
+/// @ingroup QkDag
+/// Compose the ``other`` circuit onto the ``dag`` instance with the option of a subset of
+/// of input wires of ``other`` being mapped onto a subset of output wires of this circuit.
+///
+/// ``other`` may include a smaller or equal number of wires for each type.
+///
+/// @param dag A pointer to be composed on.
+/// @param other A pointer to the DAG to compose with ``dag``.
+/// @param qubits A list of indices representing the qubit wires to compose
+///     onto.
+/// @param num_qubits The number of qubits in ``qubits``.
+/// @param clbits A list of indices representing the clbit wires to compose
+///     onto.
+/// @param num_clbits The number of qubits in ``clbits``.
+///
+/// @return an exit code.
+///
+/// # Safety
+///
+/// Behavior is undefined if ``dag`` or ``other`` are not valid, non-null pointers to a ``QkDag``.
+/// If ``qubit`` nor ``clbit`` are NULL, it must contains a less or equal amount
+/// than what the circuit owns.
+#[unsafe(no_mangle)]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_dag_compose(
+    dag: *mut DAGCircuit,
+    other: *const DAGCircuit,
+    qubits: *const u32,
+    num_qubits: usize,
+    clbits: *const u32,
+    num_clbits: usize,
+) -> ExitCode {
+    // SAFETY: Per documentation, the pointer is to valid data.
+
+    let dag = unsafe { mut_ptr_as_ref(dag) };
+    // SAFETY: Per documentation, the pointer is to valid data.
+    let other_dag = unsafe { const_ptr_as_ref(other) };
+
+    if other_dag.qubits().len() > dag.qubits().len()
+        || other_dag.clbits().len() > dag.clbits().len()
+    {
+        return ExitCode::DAGComposeMismatch;
+    }
+
+    let qubits: Option<Vec<ShareableQubit>> = if check_ptr(qubits).is_ok() {
+        let qubits: Box<[u32]> =
+            unsafe { Box::from_raw(slice_from_raw_parts(qubits, num_qubits).cast_mut()) };
+        let new_qubits: Result<Vec<ShareableQubit>, ExitCode> = qubits
+            .into_iter()
+            .map(|bit| -> Result<ShareableQubit, ExitCode> {
+                let Some(qubit) = dag.qubits().get(Qubit(bit)) else {
+                    return Err(ExitCode::DAGComposeMissingBit);
+                };
+                Ok(qubit.clone())
```

Comment:
Just a style suggestion (assuming `qubits` is a borrowed slice per the comment above):

```suggestion
        let new_qubits: Result<Vec<ShareableQubit>, ExitCode> = qubits
            .iter()
            .map(|bit| -> Result<ShareableQubit, ExitCode> {
                dag.qubits().get(Qubit(*bit)).cloned().ok_or(ExitCode::DAGComposeMissingBit)
```